### PR TITLE
flat-manager: Add 'end-of-life' and 'end-of-life-rebase' inputs

### PR DESCRIPTION
**Pull request by meisenzahl**
Related to https://github.com/elementary/appcenter-reviews/issues/293

TL;DR

> There may come a point where an application is no longer maintained. In order to inform users at update or install time that it will no longer get updates.
> 
>If the application has been renamed, you can additionally include end-of-life-rebase with the new ID. Recent flatpak versions will prompt user if they'd like to switch to the renamed app.

Source: https://github.com/flathub/flathub/wiki/App-Maintenance#end-of-life

### CODE REVIEW

File: flat-manager/index.js [hunk 1]
```diff
@@ -12,13 +12,25 @@ const run = (repository, flatManagerUrl, token) => {
...
     .then(async () => {
       let buildId = ''
-      const exitCode = await exec.exec('flat-manager-client', [
+      let args = [
         '--token',
-        token,
+        token        
+      ]
+
+      if (end_of_life && end_of_life_rebase) {
```
Comment (bilelmoussaoui) [inline flat-manager/index.js]: you can have an end-of-life without having the later but not the opposite iirc.

Comment (bilelmoussaoui) [reply flat-manager/index.js]: never mind, i was looking at an older commit

File: flat-manager/index.js [hunk 2]
```diff
@@ -61,5 +76,7 @@ if (require.main === module) {
     core.getInput('repository'),
     core.getInput('flat-manager-url'),
     core.getInput('token')
+    core.getInput('end-of-life')
```
Comment (bilelmoussaoui) [inline flat-manager/index.js:79]: missing , & on the line above

**Comment by bilelmoussaoui** [comment]
Also you need to rebuild the action so that it works, see https://github.com/bilelmoussaoui/flatpak-github-actions/blob/master/CONTRIBUTING.md#prerequisites

File: flat-manager/index.js [hunk 3]
```diff
@@ -12,13 +12,28 @@ const run = (repository, flatManagerUrl, token) => {
...
+      ]
+
+      if (end_of_life) {
+        args.push_back(`--end-of-life=${end_of_life}`)
+
+        if (end_of_life_rebase) {
+          args.push_back(`--end-of-life-rebase=${end_of_life_rebase}`)
+        } else {
+          throw Error('end-of-life has to be set if you want to use end-of-life-rebase')
+        }
```
Comment (bilelmoussaoui) [inline flat-manager/index.js:27]: This condition is wrong no? you want to throw the error if !end_of_life && end_of_life_rebase.

File: flat-manager/action.yml [hunk 4]
```diff
@@ -17,6 +17,14 @@ inputs:
     description: >
       A flat-manager token to publish into the repository.
     required: true
+  end-of-life:
```
Comment (bilelmoussaoui) [inline flat-manager/action.yml:20]: Those new options should be added to the README as well

File: flat-manager/index.js (hunk 1 above)
Comment (meisenzahl) [reply flat-manager/index.js]: Was perhaps a little too early in the morning :sweat_smile:

File: flat-manager/index.js (hunk 3 above)
Comment (meisenzahl) [reply flat-manager/index.js]: Was perhaps a little too early in the morning :sweat_smile:

**Comment by meisenzahl** [comment]
@bilelmoussaoui thank you for your feedback! I hope I have taken all the comments into account.

File: README.md [hunk 5]
```diff
@@ -124,6 +124,8 @@ jobs:
         repository: elementary
         flat-manager-url: https://flatpak-api.elementary.io
         token: some_very_hidden_token
+        end-of-life: "The application has been renamed to..."
+        end-of-life-rebase: "org.gnome.zbrown.Palette=org.gnome.zbrown.NewPalette"
```
Comment (bilelmoussaoui) [inline README.md:128]: I think we should just put the new app-id here, why do we need the old app-id?

File: README.md [hunk 6]
```diff
@@ -133,6 +135,8 @@ jobs:
 | `repository` | The repository to push the build into  | Required | - |
 | `flat-manager-url` | The flat-manager remote URL  | Required | - |
 | `token` | A flat-manager token  | Required | - |
+| `end-of-life` | Reason for end of life  | Optional | - |
+| `end-of-life-rebase` | A replacement for the repo  | Optional | - |
```
Comment (bilelmoussaoui) [inline README.md:139]: Not so sure about "A replacement of the repo", it is the new app-id isn't it?

File: README.md (hunk 5 above)
Comment (meisenzahl) [reply README.md:128]: I assume that `flat-manager-client` expects this format:

https://github.com/flatpak/flat-manager/blob/f2c1f719b35cb48b5d2b7282b66c8a7e166147b7/src/jobs.rs#L377

Comment (meisenzahl) [reply README.md:128]: If I look at the line longer

```rust
Some(format!("--end-of-life-rebase={}={}", app_ref.ref_name.split('/').nth(1).unwrap(), endoflife_rebase))
```

probably only `end_of_life_rebase` is read from the command line...

File: flat-manager/action.yml [hunk 7]
```diff
@@ -17,6 +17,14 @@ inputs:
...
+    description: >
+      Mark build as end-of-life.
+    required: false
+  end-of-life-rebase:
+    description: >
+      Mark new refs as end-of-life. This one takes an ID that supersedes the current one. By the user's request, the application data may be preserved for the new application. Note, this is actually a prefix match, so if you say org.the.app=org.new.app, then something like org.the.app.Locale will be rebased to org.new.app.Locale.
```
Comment (bilelmoussaoui) [inline flat-manager/action.yml:26]: This needs to be fixed as well

**Review by bilelmoussaoui — CHANGES_REQUESTED** [review]
Also please make the lint job happy

File: flat-manager/index.js [hunk 8]
```diff
@@ -12,13 +12,30 @@ const run = (repository, flatManagerUrl, token) => {
...
     .then(async () => {
       let buildId = ''
-      const exitCode = await exec.exec('flat-manager-client', [
+      let args = [
         '--token',
-        token,
+        token  
+      ]
+
+      if (endOfLive) {
```
Comment (bilelmoussaoui) [inline flat-manager/index.js:20]: Probably should be endOfLife. You also forgot to pass these parameters to the run function above

Comment (meisenzahl) [reply flat-manager/index.js]: I'm very sorry that I overlook so much 🙈

**Comment by bilelmoussaoui** [comment]
Thank you for your contribution! there are still some minor things I would like to see changed in the code but I will handle those myself

**Comment by meisenzahl** [comment]
@bilelmoussaoui thank you for your feedback and your patience with me.

**Pull request merged**